### PR TITLE
iouloss for yolov3

### PR DESCRIPTION
**Pull request by CodesFarmer**
Add IoU loss for yolov3

### CODE REVIEW

File: ppdet/modeling/losses/yolo_loss.py [hunk 1]
```diff
@@ -45,6 +48,8 @@ def __init__(self,
         self._ignore_thresh = ignore_thresh
         self._label_smooth = label_smooth
         self._use_fine_grained_loss = use_fine_grained_loss
+        self._MAX_WI = 608
+        self._MAX_HI = 608
```
Comment (qingqing01) [inline ppdet/modeling/losses/yolo_loss.py]: MAX_WI/HI is based on the transofrm? if so do not fixed here

File: ppdet/modeling/losses/yolo_loss.py [hunk 2]
```diff
@@ -104,15 +109,17 @@ def _get_fine_grained_loss(self, outputs, targets, gt_box, batch_size,
...
         downsample = 32
-        loss_xys, loss_whs, loss_objs, loss_clss = [], [], [], []
+        loss_xys, loss_whs, loss_locs, loss_objs, loss_clss = [], [], [], [], []
         for i, (output, target,
                 anchors) in enumerate(zip(outputs, targets, mask_anchors)):
             an_num = len(anchors) // 2
             x, y, w, h, obj, cls = self._split_output(output, an_num,
                                                       num_classes)
             tx, ty, tw, th, tscale, tobj, tcls = self._split_target(target)
+            loss_giou = self._IoUloss(x, y, w, h, tx, ty, tw, th, anchors, downsample)
```
Comment (qingqing01) [inline ppdet/modeling/losses/yolo_loss.py]: Whether need iou-loss should based the setting in the config

File: ppdet/modeling/losses/yolo_loss.py (hunk 1 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: Yeah, it is based on the transform, but I need it to create a map of anchors and then crop it for fitting the output shape of yolov3, only in this way can it support the random shape input.
Or there are better way to realize this?

File: ppdet/modeling/losses/yolo_loss.py (hunk 2 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: It can not be configured in the config, I can modify the code to support it

File: ppdet/modeling/losses/yolo_loss.py [hunk 3]
```diff
@@ -287,3 +298,135 @@ def box_xywh2xyxy(box):
...
+        y2 = pred_ctr_y + 0.5 * pred_h
+        if is_gt:
+            x1.stop_gradient = True
+            y1.stop_gradient = True
+            x2.stop_gradient = True
+            y2.stop_gradient = True
+
+        return x1, y1, x2, y2
+
+    def _IoUloss(self, x, y, w, h, tx, ty, tw, th,
```
Comment (littletomatodonkey) [inline ppdet/modeling/losses/yolo_loss.py]: 可以考虑参考
https://github.com/PaddlePaddle/PaddleDetection/blob/master/ppdet/modeling/losses/giou_loss.py
的写法，写成类的形式？方便其他模型使用~

File: ppdet/modeling/losses/yolo_loss.py [hunk 4]
```diff
@@ -287,3 +298,135 @@ def box_xywh2xyxy(box):
...
+                 anchors, downsample_ratio):
+        '''
+        IoU loss referenced the paper https://arxiv.org/abs/1908.03851
```
Comment (littletomatodonkey) [inline ppdet/modeling/losses/yolo_loss.py]: 每个参数都写下注释吧~

File: ppdet/modeling/losses/yolo_loss.py [hunk 5]
```diff
@@ -287,3 +298,135 @@ def box_xywh2xyxy(box):
...
+        yc1 = fluid.layers.elementwise_min(y1, y1g)
+        xc2 = fluid.layers.elementwise_max(x2, x2g)
+        yc2 = fluid.layers.elementwise_max(y2, y2g)
+
+        intsctk = (xkis2 - xkis1) * (ykis2 - ykis1)
+        intsctk = intsctk * fluid.layers.greater_than(
+            xkis2, xkis1) * fluid.layers.greater_than(ykis2, ykis1)
+        unionk = (x2 - x1) * (y2 - y1) + (x2g - x1g) * (y2g - y1g) - intsctk + eps
+        iouk = intsctk / unionk
+        loss_iou = 1. - iouk * iouk
```
Comment (littletomatodonkey) [inline ppdet/modeling/losses/yolo_loss.py]: 我看论文中好像是`1 - iouk`？
image

File: ppdet/modeling/losses/yolo_loss.py (hunk 2 above)
Comment (littletomatodonkey) [inline ppdet/modeling/losses/yolo_loss.py]: loss_giou -> loss_iou ?

File: ppdet/modeling/losses/yolo_loss.py (hunk 1 above)
Comment (heavengate) [reply ppdet/modeling/losses/yolo_loss.py]: 看下文只是需要获取特征图的H和W吧，从output.shape的2，3两维获取就可以了，这里最好不要硬编码

File: ppdet/modeling/losses/yolo_loss.py [hunk 6]
```diff
@@ -75,7 +80,7 @@ def __call__(self, outputs, gt_box, gt_label, gt_score, targets, anchors,
             return {'loss': sum(losses)}
 
     def _get_fine_grained_loss(self, outputs, targets, gt_box, batch_size,
-                               num_classes, mask_anchors, ignore_thresh):
+                               num_classes, mask_anchors, ignore_thresh, loss_weight=2.5):
```
Comment (heavengate) [inline ppdet/modeling/losses/yolo_loss.py]: loss_weight  -> iou_loss_weight ?

File: ppdet/modeling/losses/yolo_loss.py [hunk 7]
```diff
@@ -104,15 +109,17 @@ def _get_fine_grained_loss(self, outputs, targets, gt_box, batch_size,
             "YOLOv3 output layer number not equal target number"
 
         downsample = 32
-        loss_xys, loss_whs, loss_objs, loss_clss = [], [], [], []
+        loss_xys, loss_whs, loss_locs, loss_objs, loss_clss = [], [], [], [], []
```
Comment (heavengate) [inline ppdet/modeling/losses/yolo_loss.py]: Yolo的paper里面一般称(x, y, w, h)为location loss，这里 loss_locs改成loss_ious ?

File: ppdet/modeling/losses/yolo_loss.py [hunk 8]
```diff
@@ -287,3 +298,135 @@ def box_xywh2xyxy(box):
...
+        grid_y = int(self._MAX_HI / downsample_ratio)
+
+        shape_fmp = fluid.layers.shape(dcx)
+        shape_fmp.stop_gradient = True
+        # generate the grid_w x _grid_h center of feature map
+        idx_i = np.array([[i for i in range(grid_x)]])
+        idx_j = np.array([[j for j in range(grid_y)]]).transpose()
+        gi_np = np.repeat(idx_i, grid_y, axis=0)
+        gi_np = np.expand_dims(gi_np, axis=0)
+        gi_np = np.expand_dims(gi_np, axis=0)
```
Comment (heavengate) [inline ppdet/modeling/losses/yolo_loss.py]: 这两行换成reshape，两个expand_dims有点绕？

File: ppdet/modeling/losses/yolo_loss.py [hunk 9]
```diff
@@ -287,3 +298,135 @@ def box_xywh2xyxy(box):
...
+        gi_np = np.repeat(gi_np, 3, axis=1)
+        gi_np = np.repeat(gi_np, batch_size, axis=0)
```
Comment (heavengate) [inline ppdet/modeling/losses/yolo_loss.py]: 两个repeat可以改成tile ？

File: ppdet/modeling/losses/yolo_loss.py [hunk 10]
```diff
@@ -287,3 +298,135 @@ def box_xywh2xyxy(box):
...
+        gj_np = np.repeat(idx_j, grid_x, axis=1)
+        gj_np = np.expand_dims(gj_np, axis=0)
+        gj_np = np.expand_dims(gj_np, axis=0)
+        gj_np = np.repeat(gj_np, 3, axis=1)
+        gj_np = np.repeat(gj_np, batch_size, axis=0)
+        gi_max = self._crate_tensor_from_numpy(gi_np.astype(np.float32))
```
Comment (heavengate) [inline ppdet/modeling/losses/yolo_loss.py]: crate -> create ?

File: ppdet/modeling/losses/yolo_loss.py [hunk 11]
```diff
@@ -287,3 +298,135 @@ def box_xywh2xyxy(box):
...
+
+        return x1, y1, x2, y2
+
+    def _IoUloss(self, x, y, w, h, tx, ty, tw, th,
+                 anchors, downsample_ratio):
+        '''
+        IoU loss referenced the paper https://arxiv.org/abs/1908.03851
+        Loss = 1.0 - iou * iou
+        '''
+        eps = 1.e-10
```
Comment (heavengate) [inline ppdet/modeling/losses/yolo_loss.py]: 这里eps作为这个函数的一个默认参数吧

File: ppdet/modeling/losses/yolo_loss.py [hunk 12]
```diff
@@ -287,3 +298,135 @@ def box_xywh2xyxy(box):
...
+        intsctk = (xkis2 - xkis1) * (ykis2 - ykis1)
+        intsctk = intsctk * fluid.layers.greater_than(
+            xkis2, xkis1) * fluid.layers.greater_than(ykis2, ykis1)
+        unionk = (x2 - x1) * (y2 - y1) + (x2g - x1g) * (y2g - y1g) - intsctk + eps
+        iouk = intsctk / unionk
+        loss_iou = 1. - iouk * iouk
+
+        return loss_iou
+
+    def _crate_tensor_from_numpy(self, numpy_array):
```
Comment (heavengate) [inline ppdet/modeling/losses/yolo_loss.py]: create？

File: ppdet/modeling/losses/yolo_loss.py (hunk 3 above)
Comment (heavengate) [inline ppdet/modeling/losses/yolo_loss.py]: _iou_loss ?

File: ppdet/modeling/losses/yolo_loss.py (hunk 10 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: 这个命名我改一下

File: ppdet/modeling/losses/yolo_loss.py (hunk 9 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: 好的

File: ppdet/modeling/losses/yolo_loss.py (hunk 5 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: 实际上我参考的是这个实现

Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: 实际上我参考的是这个实现

Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: 我实际参考的是这个实现https://github.com/ruinmessi/ASFF/blob/master/models/utils_loss.py#L60 最后实验结果也是按照这个实现验证有效的

File: ppdet/modeling/losses/yolo_loss.py (hunk 1 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: paddle有没有什么函数可以动态的根据别的tensor的shape构建一个anchor map？我之所以这样写是为了从numpy array中构建这个anchor map，以便支持随机分辨率的输入

File: ppdet/modeling/losses/yolo_loss.py (hunk 8 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: 好的

File: ppdet/modeling/losses/yolo_loss.py (hunk 3 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: OK

File: ppdet/modeling/losses/yolo_loss.py (hunk 12 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: OK

File: ppdet/modeling/losses/yolo_loss.py (hunk 7 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: OK

File: ppdet/modeling/losses/yolo_loss.py (hunk 6 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: OK

File: ppdet/modeling/losses/yolo_loss.py (hunk 3 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: yolo的输出还需要再decode一下，和faster-rcnn输出不太一样，如果要改transform函数需要不同实现

File: ppdet/modeling/losses/yolo_loss.py (hunk 4 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: OK

File: ppdet/modeling/losses/yolo_loss.py (hunk 2 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: OK

File: ppdet/modeling/losses/yolo_loss.py (hunk 11 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/yolo_loss.py]: ok

**Comment by CodesFarmer** [comment]
> 最好独立出去作为可配置模块，另外需要随pr合入配置文件，model zoo模型这些

好的

**Comment by littletomatodonkey** [comment]
model zoo中也添加相应的模型吧

File: ppdet/modeling/losses/iou_loss.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+        intsctk = intsctk * fluid.layers.greater_than(
+            xkis2, xkis1) * fluid.layers.greater_than(ykis2, ykis1)
+        unionk = (x2 - x1) * (y2 - y1) + (x2g - x1g) * (y2g - y1g) - intsctk + eps
+        iouk = intsctk / unionk
+        loss_iou = 1. - iouk * iouk
+        loss_iou = loss_iou * self._loss_weight
+
+        return loss_iou
+
+    def _bbox_transform(self, dcx, dcy, dw, dh, anchors, downsample_ratio, batch_size, is_gt):
```
Comment (heavengate) [inline ppdet/modeling/losses/iou_loss.py]: 这里应该可以用fluid.layers.yolo_box ?

File: configs/dcn/yolov3_reader.yml [hunk 14]
```diff
@@ -0,0 +1,104 @@
+TrainReader:
```
[the remark this answers is not in the document]
Comment (CodesFarmer) [reply configs/dcn/yolov3_reader.yml]: ok

File: ppdet/modeling/losses/iou_loss.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
+# Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (CodesFarmer) [reply ppdet/modeling/losses/iou_loss.py]: ok

File: ppdet/modeling/losses/iou_loss.py [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+from ppdet.core.workspace import register, serializable
+
+__all__ = ['IouLoss']
+
+
+@register
+@serializable
+class IouLoss(object):
+    """
+    Combined loss for YOLOv3 network
```
[the remark this answers is not in the document]
Comment (CodesFarmer) [reply ppdet/modeling/losses/iou_loss.py]: OK

File: ppdet/modeling/losses/iou_loss.py (hunk 13 above)
Comment (CodesFarmer) [reply ppdet/modeling/losses/iou_loss.py]: yolo_box当时考虑过，iou loss计算的时候需要知道每个输出结果对应的gt，这个无法通过gt_box, gt_label, gt_score直接获得，另外也不能通过yolo_box来对targets直接进行解码，用这个实现也比较麻烦

File: ppdet/modeling/losses/yolo_loss.py [hunk 17]
```diff
@@ -34,17 +34,20 @@ class YOLOv3Loss(object):
...
                  ignore_thresh=0.7,
                  label_smooth=True,
-                 use_fine_grained_loss=False):
+                 use_fine_grained_loss=False,
+                 iou_loss=None):
         self._batch_size = batch_size
         self._ignore_thresh = ignore_thresh
         self._label_smooth = label_smooth
         self._use_fine_grained_loss = use_fine_grained_loss
+        self._iou_loss = iou_loss
```
Comment (qingqing01) [inline ppdet/modeling/losses/yolo_loss.py]: 这里需要判断下，iou_loss非None的时候，use_fine_grained_loss必需为True
